### ISSUE TITLE
Benchmarking documentation

### ISSUE THREAD

**Issue by fabiannagel** [google/jax-md#126]
Hi! I'm currently working on a project to evaluate whether porting our existing MD infrastructure to JAX-MD is a viable option for us. As the focus is mostly runtime performance, I was looking for documentation on how to properly do this. 

I'm sure you know that plain [JAX](https://jax.readthedocs.io/en/latest/faq.html#benchmarking-jax-code) has some really nice info on this. I'd be interested to find out how these approaches translate to JAX-MD, i.e.
- Is there a way to passthrough `block_until_ready()`? 
- Are there implications for manually using `device_put()`? 

Also, I remember that you had some performance benchmarks yourself in the first JAX-MD paper. I'd be very interested to find out more about the implementational details on this. I provide some context below on what I'm trying to do, but I'd appreciate any pointer in the right direction. Thanks for the great work :)

--------------

I've been investigating the performance of JAX-MD's Lennard-Jones calculator for constant density systems of increasing size. This is partly to determine the real-life computational demands of stress computations (see #116). We perform a warm-up to get `jit` compilation costs out of the equation, followed by 100 runs of LJ energy computations of the same system. All of this is repeated with increasing atom counts. Runtimes are measured by using Python's `time.monotonic()` clock and computation results are dispatched to plain `numpy` arrays to enforce synchronicity. 

The plot below shows the runtime (y axes) plotted over its indices of those 100 runs (x axes), broken down per atom count.
In general, I would expect less fluctuation, at least for large system in which the ratio of overhead to computational cost is smaller. I find the bottom 4 plots particularly puzzling since the initial (although already compiled) calls seem to be the cheapest ones. For the potential case of missed (additional) compilation or caching mechanisms, I would expect a consistent behavior like in plot `n = 32` (top right).

<img width="1138" alt="Screenshot 2021-03-10 at 00 42 06" src="https://user-images.githubusercontent.com/4445568/110554117-74a52b80-813a-11eb-9f39-5d1df0f0a094.png">

**Comment by sschoenholz** [comment]
Hey Fabian,

Great questions and thanks for looking into JAX-MD! I agree that we should have a public benchmark available and write up some best practices we've found around timing / writing fast code. To that end, [here](https://colab.research.google.com/github/google/jax-md/blob/master/notebooks/lj_benchmark.ipynb) is a copy of the benchmarks from the paper to get started (added in 13202f73509decd29f98322344793b50381c6174). I hope to add another benchmark comparing jit vs not-jit execution. In the meantime I sort of go through an example in this [notebook](https://colab.research.google.com/github/google/jax-md/blob/master/notebooks/talk_demo.ipynb).

I also agree that the timing you're seeing looks a bit strange, though what you are doing sounds correct. One note is that if you dispatch to `numpy` arrays to synchronize then you will also time the device transfer (which will not happen with `block_until_ready()` AFAIK). Do you think you could send me the code used to generate the plots so I can check whether there is an obvious place where the JIT cache is being missed (which is the most common source of occasional increases in runtime)?

A few observations general observations.

1. We have observed that, especially for smaller systems, it can be very beneficial to JIT together multiple steps. In general, I usually develop simulations without trying to JIT everything, but then convert the code to so that large blocks are JIT before running large experiments / benchmarks. Even with the asynchronous dispatch, I have found interaction with Python can add significantly to the runtime. Also, jitting multiple steps together allows XLA to do better optimization. Usually simulations will end up look something like this:
    ```python
    neighbor_fn, energy_fn = energy.lennard_jones_neighbor_list(displacement, box_size)
    init, apply = simulate.nvt_nose_hoover(energy_fn, shift, 5e-3, kT=1.2)
    def step(i, state_and_nbrs):
      state, nbrs = state_and_nbrs
      nbrs = neighbor_fn(state.position, nbrs)
      return apply(state, neighbor=nbrs), nbrs

    steps = 10000
    new_state, new_nbrs = lax.fori_loop(0, steps, step, (state, nbrs))
    if new_nbrs.did_buffer_overflow:
      nbrs = neighbor_fn(state.position)
    else:
      state = new_state
      nbrs = new_nbrs
    ```
    When doing benchmarking, I usually time the code for some relatively large number of steps.
2. Especially for large systems, we see a very significant benefit to using neighbor lists.
3. I used to worry a lot about using `device_put()` to manually shuffle data to / from device. I have not been doing it lately since for my use-cases data typically stays on device (and so the initial transfer is not a big deal). If you are experiencing OOM issues or otherwise need to have more control over memory, then I don't think there are any real gotchas here.
4. I'm not totally sure I understand what you mean by "Is there a way to passthrough block_until_ready()?", can you clarify what you're going for?

If you have any more questions, definitely feel free to ask. I think discussing benchmarking / speed of JAX-MD is always super useful.

**Comment by fabiannagel** [comment]
Hey Sam, thanks for your detailed reply. 

1. I've been looking for a while for some "jit best practises", although I just saw there is a new [JAX 101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/index.html) with some more info on this. Nonetheless, your help is much appreciated!

There is definitely room for improvement in terms of `jit` handling in my [code](https://github.com/fabiannagel/gknet-benchmarks/blob/master/calculators/lennard_jones/pair/jaxmd_lennard_jones_pair.py#L84). I'll try to give it some polish over the next few days and incorporate your ideas for improvement. Maybe this will already solve those strange timings.

Some background info on the repository might be useful for you: Basically, we're trying to figure out whether using `ASE` as a MD framework with a GPU-based `JAX-MD` Lennard-Jones calculator solves our performance issues. If not, we want to investigate whether switching to end-to-end `JAX-MD` would be worth the effort.

2. Neighbor lists are definitely within our scope. So far I've been hesitant to benchmark them in detail because of their additional requirements for `jit` compilation. If LJ pair looks fine w.r.t. the described runtime variance, I will definitely include them!

3. OOM issues are in fact something I have experienced surprisingly early for simple LJ systems. Around 20k atoms on a P100 (16 GB VRAM) if I remember correctly (energy/forces only, no large Jacobians). I tried this a few months ago when I had first gotten into JAX and JAX-MD, so maybe I missed some beginner's pitfalls. 

4. To be honest, I might have overthought the interplay between JAX and JAX-MD. After all, `block_until_ready()` is simply a method on every `jax.numpy` array that is also returned by JAX-MD.

**Comment by sschoenholz** [comment]
Hi Fabian,

Sorry for the delay in response, last week was a bit busy with ICML rebuttals. I think regarding 3., the OOM you are seeing might be correct without neighbor lists since JAX MD instantiates a full 20k x 20k x 3 matrix to compute the energy.

If it would be helpful, perhaps we can schedule a short meeting to discuss this in a higher bandwidth setting? Overall, I'm super interested in the question of how well JAX MD works as a calculator for ASE. It seems like you've already done some really useful exploration in that direction and I'd love to help it work well.

Best,
Sam

**Comment by fabiannagel** [comment]
Hi Sam,

sorry for the late response also from my side! Somehow I missed the notification. Absolutely, a meeting would be very interesting for us as well. These suggested dates (Pacific Daylight Time) would work for us:

- May 5, 10 - 11 am
- May 6, 07 - 11 am
- May 7, 07 - 11 am

Also, please let me know if there's a better tool to schedule meetings across timezones like this ;) In any case, feel free to shoot me an [email](mailto:fabian.2.nagel@uni.kn) for further discussion. Looking forward to your reply!

Best,
Fabian

**Comment by fabiannagel** [comment]
Hi Sam, just wanted to give you a quick heads-up on everything. Fortunately, the issues with runtime variance are resolved. I accidentally dispatched `DeviceArrays` to a Python dataclass with `numpy.array` members, but that's obviously more of a type hint rather than explicit construction of `numpy` arrays... explicitly calling `block_until_ready()` did the trick for me.

Ultimately, I created some runtime benchmarks for multiple potentials (LJ pair, LJ NL, `energy.graph_network_neighbor_list`)
for constant-density systems of increasing size. We evaluate different property sets to investigate the impact of computing stress(-es) on runtime and memory requirements. As a next step, we're currently building an `ASE`-compatible LJ NL calculator to benchmark NVE performance in a hybrid ASE-JAX-MD setting versus NVE in plain JAX-MD.

We're still very interested in getting together to discuss details. Please let me know what time and date would fit your schedule :)

Best,
Fabian

**Comment by sschoenholz** [comment]
Hi Fabian,

That sounds like fantastic progress, thanks for doing this work! Sorry that I somehow missed your last reply; lately I have been overflowing with email. Let's definitely meet, I'm so curious to see your benchmarks and use them to figure out ways jax md could be faster.

I'll reach out by email to coordinate further.

Best,
Sam
